src/aleph/tcp.clj

```diff
@@ -40,15 +52,13 @@
 
       :channel-active
       ([_ ctx]
-        (let [ch (.channel ctx)]
-          (handler
-            (doto
-              (s/splice
-                (netty/sink ch true netty/to-byte-buf)
-                (reset! in (netty/source ch)))
-              (reset-meta! {:aleph/channel ch}))
-            (->TcpConnection ch)))
-        (.fireChannelActive ctx))
+       (if-let [^SslHandler ssl-handler (-> ctx .pipeline (.get SslHandler))]
+         (-> ssl-handler
+             .handshakeFuture
+             netty/wrap-future
+             (d/chain' (fn [_] (call-handler ctx))))
```

Comment:
@arnaudgeiser @DerGuteMoritz 

We have a slight bug here.

First, Moritz, if you don't need the deferred's value, you probably want to use `on-realized`, and not `chain/chain'`.

Second, it's still not checking for success here. Or rather, `wrap-future` IS, but deceptively, which brings me to the problem. Technically, it's not a bug, since the code works, but it _is_ confusing and wasteful. `wrap-future` uses `.isSuccess` in a way where it really wants to call `.isDone`. 

If the future succeeded, it immediately makes a success-deferred. If it failed, it attaches `operation-complete` as a listener. Note that if the future was cancelled or completed with an exception, it will then call `operation-complete` immediately, which may return an error-deferred. It's wasted work, since it may set up `operation-complete` to run on another thread, though.

More importantly, the call to `.isSuccess` isn't guarding anything for you, since you could still get an error-deferred. `on-realized` will force you to handle both success and error scenarios.

We should probably change `wrap-future` to use `isDone`, and immediately invoke `operation-complete` if true. But not in this PR.

Reply:
Oof, tricky business! But I think I successfully applied `wrap-head` to your comment now :sweat_smile: Changing `wrap-future` the way you suggest indeed sounds like a reasonable improvement :+1:

Changing to `d/on-realized` is a good idea, too. Not only does it make the intent clearer, but I think my current implementation would also generate extra overhead if the handler happens to return a deferred. My only question is what to do with the error handler. As you noticed, the code already works without explicit handling of the error because the SSL handler terminates the pipeline in this case anyway (by throwing a `javax.net.ssl.SSLHandshakeException`). I'm passing a (commented) noop function now, just like `d/listener` does in the single arity case. Lemme know if you have a different understanding or better idea!

Reply:
I think we should log it as a warning.

Reply:
This is actually already covered via the general exception handler. Results in a log like this:

```
[aleph-netty-server-event-pool-2] WARN aleph.http.server - error in HTTP server
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: Invalid CertificateVerify signature
[ ... stack trace omitted ... ]
```

See also https://github.com/clj-commons/aleph/pull/611

In other words: Logging it here, too, would result in a redundant warning for the same handshake failure. Would you still prefer to have it?